proof-systems/src/darlin/tests/simple_marlin.rs

```diff
@@ -0,0 +1,161 @@
+};
+use crate::darlin::pcd::{
+    PCDParameters, simple_marlin::SimpleMarlinPCD
+};
+use rand::{ Rng, RngCore };
+use digest::Digest;
+use std::ops::MulAssign;
+
+/// A simple test circuit with two field elements c,d as inputs, enforced to satisfy
+///     (c,d) = a*(b,b^2),
+/// To produce the given `num_constraints`, the same two constraints
+///     a * b = c
+///     c * b = d
+/// are repeated accordingly. To acchieve the give `num_variables`, a corresponding number of
+/// dummy witness variables are allocated.
+#[derive(Copy, Clone)]
+pub(crate) struct Circuit<F: Field> {
+    a: Option<F>,
+    b: Option<F>,
+    num_constraints: usize,
+    num_variables: usize,
+}
+
+impl<ConstraintF: Field> ConstraintSynthesizer<ConstraintF> for Circuit<ConstraintF> {
+    fn generate_constraints<CS: ConstraintSystem<ConstraintF>>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(), SynthesisError> {
+        let a = cs.alloc(|| "a", || self.a.ok_or(SynthesisError::AssignmentMissing))?;
+        let b = cs.alloc(|| "b", || self.b.ok_or(SynthesisError::AssignmentMissing))?;
+        let c = cs.alloc_input(
+            || "c",
+            || {
+                let mut a = self.a.ok_or(SynthesisError::AssignmentMissing)?;
+                let b = self.b.ok_or(SynthesisError::AssignmentMissing)?;
+
+                a.mul_assign(&b);
+                Ok(a)
+            },
+        )?;
+        let d = cs.alloc_input(
+            || "d",
+            || {
+                let mut a = self.a.ok_or(SynthesisError::AssignmentMissing)?;
+                let b = self.b.ok_or(SynthesisError::AssignmentMissing)?;
+
+                a.mul_assign(&b);
+                a.mul_assign(&b);
+                Ok(a)
+            },
+        )?;
+
+        for i in 0..(self.num_variables - 5) {
+            let _ = cs.alloc(
+                || format!("var {}", i),
+                || self.a.ok_or(SynthesisError::AssignmentMissing),
+            )?;
+        }
+
+        for i in 0..(self.num_constraints - 1){
+            cs.enforce(
+                || format!("constraint {}", i),
+                |lc| lc + a,
+                |lc| lc + b,
+                |lc| lc + c,
+            );
+        }
+        cs.enforce(
+            || format!("constraint {}", self.num_constraints - 1),
+            |lc| lc + c,
+            |lc| lc + b,
+            |lc| lc + d,
+        );
+        Ok(())
+    }
+}
+
+/// Generates a SimpleMarlinPCD from `Circuit`, by sampling the internal
+/// witnesses a,b at random.
+#[allow(dead_code)]
+pub(crate) fn generate_test_pcd<'a, G: AffineCurve, D: Digest + 'a, R: RngCore>(
+    pc_ck: &CommitterKey<G>,
+    marlin_pk: &MarlinProverKey<G::ScalarField, InnerProductArgPC<G, D>>,
+    num_constraints: usize,
+    zk: bool,
+    rng: &mut R,
+) -> SimpleMarlinPCD<'a, G, D>
+{
+    let a = G::ScalarField::rand(rng);
+    let b = G::ScalarField::rand(rng);
+    let mut c = a;
+    c.mul_assign(&b);
+    let mut d = c;
+    d.mul_assign(&b);
```

Comment:
Again a synthesizer related question: why do we compute the public inputs separately and not let it be returned by the synthesizer?

Reply:
As above: with the functions and interfaces we have know we cannot let the synthesizer "return" anything

Reply:
As above, let's put this on our list.